packages/loopring_v3/contracts/impl/ExchangeV3.sol

```diff
@@ -519,7 +558,87 @@ contract ExchangeV3 is IExchangeV3
         payable
         nonReentrant
     {
-        state.depositTo(recipient, tokenAddress, amount, 0);
+        state.deposit(msg.sender, recipient, tokenAddress, amount, 0);
+    }
+
+    function deposit(
+        address from,
+        address to,
+        address tokenAddress,
+        uint96  amount
+        )
+        external
+        payable
+        nonReentrant
+    {
+        state.deposit(from, to, tokenAddress, amount, 0);
+    }
+
+    function approveConditionalTransfer(
+        address from,
+        address to,
+        address token,
+        uint24  fAmount,
+        address feeToken,
+        uint24  fFee,
+        uint32  salt
+        )
+        external
+        nonReentrant
+        onlyRelayer
+    {
+        uint24 fromAccountID = state.getAccountID(from);
+        uint24 toAccountID = state.getAccountID(from);
+        uint16 tokenID = state.getTokenID(token);
+        uint16 feeTokenID = state.getTokenID(feeToken);
+
+        // Same packing as onchain DA + salt
+        uint key = fromAccountID;
+        key <<= 20;
+        key |= toAccountID;
+        key <<= 8;
+        key |= tokenID;
+        key <<= 24;
+        key |= fAmount;
+        key <<= 8;
+        key |= feeTokenID;
+        key <<= 16;
+        key |= fFee;
+        key <<= 8;
+        key |= 1;
+        key <<= 32;
+        key |= salt;
+
+        // Make sure we didn't already approve this conditional transfer
+        require(!state.conditionalTransfers[key].approved, "TRANSFER_ALREADY_APPROVED");
+        state.conditionalTransfers[key] = ExchangeData.ConditionalTransferState(
+            true,
+            0,
+            0
+        );
+
+        emit ConditionalTransferApproved(
+            from,
+            to,
+            token,
+            uint96(uint(fAmount).decodeFloat(24)),
+            feeToken,
+            uint96(uint(fFee).decodeFloat(16)),
+            salt
+        );
+    }
+
+    function onchainTransferFrom(
+        address from,
+        address to,
+        address token,
+        uint    amount
+        )
+        external
+        nonReentrant
+        onlyRelayer
+    {
+        require(token.safeTransferFrom(from, to, amount), "TRANSFER_FAILED");
```

Comment:
so the point of introducing this `onchainTransferFrom` function is to avoid another user approval? Since this method doesn't have any condition checking, it means the caller, or the Relayer contract cannot be signed dynamically and the address must be provided during DEX initialization to reduce the risk (otherwise if the relayer contract can be set later, then the operator can still people's token)

Reply:
Correct. The Relayer would be something that is also exclusively handled by Loopring and we could perhaps make it updatable using the protocol contract, though because the main Relayer contract doesn't do much (just forwards call to modules) I don't think that's even needed. We could make it so that exchanges could opt in or out of modules, though I also don't think that's necessary because modules will just call exchange functions that it would need to process anyway.